src/libs/ReportUtils.ts

```diff
@@ -6344,10 +6351,43 @@ function hasActionsWithErrors(reportID: string): boolean {
     return Object.values(reportActions).some((action) => !isEmptyObject(action.errors));
 }
 
-function canLeavePolicyExpenseChat(report: OnyxEntry<Report>, policy: OnyxEntry<Policy>): boolean {
+function isNonAdminOrOwnerOfPolicyExpenseChat(report: OnyxEntry<Report>, policy: OnyxEntry<Policy>): boolean {
     return isPolicyExpenseChat(report) && !(PolicyUtils.isPolicyAdmin(policy) || PolicyUtils.isPolicyOwner(policy, currentUserAccountID ?? -1) || isReportOwner(report));
 }
 
+/**
+ * Whether the user can join a report
+ */
+function canJoinChat(report: OnyxEntry<Report>, parentReportAction: OnyxEntry<ReportAction>, policy: OnyxEntry<Policy>): boolean {
+    // We disabled thread functions for whisper action
+    // So we should not show join option for existing thread on whisper message that has already been left, or manually leave it
+    if (ReportActionsUtils.isWhisperAction(parentReportAction)) {
+        return false;
+    }
+
+    // If the notification preference of the chat is not hidden that means we have already joined the chat
+    if (report?.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN) {
+        return false;
+    }
+
+    if (isRootGroupChat(report) || isSelfDM(report)) {
+        return false;
+    }
+
+    return isChatReport(report) || canLeaveRoom(report, !isEmptyObject(policy)) || isNonAdminOrOwnerOfPolicyExpenseChat(report, policy);
```

Comment:
I still think we should get rid of `canLeaveRoom()`?  It's a bit of a mess.

The process I'd take here is to analyze the current code and figure out what conditions would make a chat "joinable" and what conditions would make a chat "leaveable". Actually someone did it here:

https://github.com/Expensify/App/blame/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/libs/ReportUtils.ts#L5478-L5487

So we can go down that list and find what is also true for "join" and then add it to this method instead...

Several of the options say "Nobody can leave". Do we need to consider those cases here? Probably not, because our method is returning `true` at the end.

Reply:
I would also recommend the structure we are using here for both methods i.e. return early for any disqualifying conditions then return the qualifying conditions at the end.

Reply:
@marcaaron After checking `canLeaveRoom` function again, I think we can remove this function like this

1. For invoice room case, we can create a separate function `canJoinOrLeaveInvoiceRoom`

2. For other cases in this function, because only `POLICY_ROOM` chat type has `report.visibility` and all other chat types can't leave room as description of this function, we can earlier return false if the report is policyAnnounce room and the user the policy member.


After refactoring, we can remove `canLeaveRoom` function and `canLeaveChat` function will be like this

```
function canLeaveChat(report: OnyxEntry<Report>, policy: OnyxEntry<Policy>): boolean {
    if (isSelfDM(report) || isRootGroupChat(report)) {
        return false;
    }

    if (isPublicAnnounceRoom(report) && !isEmptyObject(policy)) {
        return false;
    }

    if (canJoinOrLeaveInvoiceRoom(report)) {
        return true;
    }

    return (isChatThread(report) && !!report?.notificationPreference?.length) || isUserCreatedPolicyRoom(report) || isNonAdminOrOwnerOfPolicyExpenseChat(report, policy);
}
```

cc @jjcoffee

Reply:
Sorry, can you just apply whatever changes you think we should apply and I will review them in the context of the greater PR review? At first glance... I'm not really loving the `canJoinOrLeaveInvoiceRoom()` here. It's not clear to me what case someone can "join" an invoice room, but makes sense that we would not allow someone to leave one.